releasing/plugin-commands-deploy/src/createDeployFiles.ts

```diff
@@ -0,0 +1,197 @@
+      })
+
+      if (!targetRealPath) {
+        targetSpecifiers[name] = targetDependencies[name] = spec
+        continue
+      }
+
+      targetSpecifiers[name] = targetDependencies[name] =
+        targetRealPath === deployedProjectRealPath ? 'link:.' : createFileUrlDepPath(targetRealPath as ProjectRootDirRealPath, allProjects)
+    }
+  }
+
+  return {
+    lockfile: {
+      ...lockfile,
+      importers: {
+        ['.' as ProjectId]: targetSnapshot,
+      },
+      packages: targetPackageSnapshots,
+    },
+    manifest: {
+      ...pick(INHERITED_MANIFEST_KEYS, manifest),
+      dependencies: targetSnapshot.dependencies,
+      devDependencies: targetSnapshot.devDependencies,
+      optionalDependencies: targetSnapshot.optionalDependencies,
+    },
+  }
+}
+
+interface ConvertOptions {
+  allProjects: CreateDeployFilesOptions['allProjects']
+  deployedProjectRealPath: ProjectRootDirRealPath
+  projectRootDirRealPath: ProjectRootDirRealPath
+  lockfileDir: string
+}
+
+function convertProjectSnapshotToPackageSnapshot (projectSnapshot: ProjectSnapshot, opts: ConvertOptions): PackageSnapshot {
+  const resolution: DirectoryResolution = {
+    type: 'directory',
+    directory: '.',
+  }
+  const dependencies = convertResolvedDependencies(projectSnapshot.dependencies, opts)
+  const optionalDependencies = convertResolvedDependencies(projectSnapshot.optionalDependencies, opts)
+  return {
+    dependencies,
+    optionalDependencies,
+    resolution,
+  }
+}
+
+function convertResolvedDependencies (input: ResolvedDependencies | undefined, opts: ConvertOptions): ResolvedDependencies | undefined {
+  if (!input) return undefined
+  const output: ResolvedDependencies = {}
+
+  for (const key in input) {
+    const spec = input[key]
+    const depRealPath = resolveLinkOrFile(spec, opts)
+    if (!depRealPath) {
+      output[key] = spec
+      continue
+    }
+
+    if (depRealPath === opts.deployedProjectRealPath) {
+      output[key] = 'link:.' // the path is relative to the lockfile dir, which means '.' would reference the deploy dir
+      continue
+    }
+
+    output[key] = createFileUrlDepPath(depRealPath as ProjectRootDirRealPath, opts.allProjects)
+  }
+
+  return output
+}
+
+function resolveLinkOrFile (spec: string, opts: Pick<ConvertOptions, 'lockfileDir' | 'projectRootDirRealPath'>): string | undefined {
+  // try parsing `spec` as either @scope/name@pref or name@pref
+  const renamed = /^@(?<scope>[^@]+)\/(?<name>[^@]+)@(?<pref>.+)$/.exec(spec) ?? /^(?<name>[^@])+@(?<pref>.+)$/.exec(spec)
+  if (renamed) return resolveLinkOrFile(renamed.groups!.pref, opts)
+
+  const { lockfileDir, projectRootDirRealPath } = opts
+
+  if (spec.startsWith('link:')) {
+    const targetPath = spec.slice('link:'.length)
+    return path.resolve(projectRootDirRealPath, targetPath)
+  }
+
+  if (spec.startsWith('file:')) {
+    const targetPath = spec.slice('file:'.length)
+    return path.resolve(lockfileDir, targetPath)
+  }
+
+  return undefined
+}
```

Comment:
I tried to find an existing function in the `resolver` directory, but none of them can fit nicely.

Also, this whole function can be converted to a single regex. Should we do that?